src/iter/par_bridge.rs

```diff
@@ -135,6 +123,25 @@ where
     where
         F: Folder<Self::Item>,
     {
+        // Guard against work-stealing-induced recursion, in case `Iter::next()`
+        // calls rayon internally, so we don't deadlock our mutex. We might also
+        // be recursing via `folder` methods, which doesn't present a mutex hazard,
+        // but it's lower overhead for us to just check this once, rather than
+        // updating additional shared state on every mutex lock/unlock.
+        // (If this isn't a rayon thread, then there's no work-stealing anyway...)
+        if let Some(i) = current_thread_index() {
+            // Note: If the number of threads in the pool ever grows dynamically, then
```

Comment:
Two ideas:

* If this potentially approximate in any case, couldn't we use e.g. an `AtomicU64` as a bitmask indicating which of the 64 groups of threads has started work on the bridge and avoid the `Vec`?
* Since we deadlock when the current thread already owns the mutex, wouldn't it suffice to store the current owner's thread ID instead of separate flags and then check if the current thread already holds the mutex? (I am admittedly unsure how to properly synchronize holding the mutex and- updating/checking the current-owner-thread-ID.)

Reply:
> (I am admittedly unsure how to properly synchronize holding the mutex and updating/checking the current-owner-thread-ID.)

But that might not be a problem since we only really care about the case of synchronizing with ourselves, right? So updating the owner ID from "none" to "me" after locking and before unlocking should suffice so that the current thread can load it and check "it is none or not me" before trying to lock, shouldn't it? (Because either we are racing with someone else and don't care or we are going to deadlock and then it is guaranteed that the atomic was updated before we got here for the second locking attempt.)